File: lobster-input-front/voice-input/Core/Dictionary/DictionaryStore.swift
/// DictionaryStore.swift
/// 热词词典数据管理（单例 ObservableObject），驱动 DictionaryView 的列表展示与 CRUD 操作。
/// 支持点击式翻页和关键词模糊搜索，无数量上限。
import Foundation
import Combine

@MainActor
final class DictionaryStore: ObservableObject {

    static let shared = DictionaryStore()
    private init() {}

    static let pageSize      = 50
    static let maxWordLength = 20

    // MARK: - 列表状态

    @Published private(set) var hotwords: [HotWordItem] = []
    @Published private(set) var isLoading = false
    @Published var errorMessage: String?

    /// 当前页码（1-indexed）
    @Published private(set) var currentPage = 1
    /// 当前搜索词对应的总条数
    @Published private(set) var total = 0
    /// 总页数
    var totalPages: Int { max(1, Int(ceil(Double(total) / Double(Self.pageSize)))) }
    var hasPrev: Bool { currentPage > 1 }
    var hasNext: Bool { currentPage < totalPages }

    // MARK: - 搜索状态

    /// 当前生效的搜索关键词（由 View 写入，Store 监听后自动触发重新加载）
    @Published var searchText: String = ""

    private var searchTask: Task<Void, Never>?
    private var loadedSearchText: String = ""

    private let api = APIClient.shared

    // MARK: - 加载

    /// 初次进入页面或手动刷新时调用，重置到第1页
    func load() async {
        currentPage = 1
        loadedSearchText = searchText
        await fetchPage(page: 1, search: searchText)
    }

    /// 点击式翻页：跳转到指定页（保持当前搜索词不变）
    func goToPage(_ page: Int) async {
        guard page >= 1, page <= totalPages else { return }
        currentPage = page
        await fetchPage(page: page, search: loadedSearchText)
    }

    func prevPage() async { await goToPage(currentPage - 1) }
    func nextPage() async { await goToPage(currentPage + 1) }

    /// 搜索词变化时调用（防抖 300ms）
    func onSearchChanged() {
        searchTask?.cancel()
        searchTask = Task {
            try? await Task.sleep(nanoseconds: 300_000_000)
            guard !Task.isCancelled else { return }
            currentPage = 1
            loadedSearchText = searchText
            await fetchPage(page: 1, search: searchText)
        }
    }

    // MARK: - 私有加载

    private func fetchPage(page: Int, search: String) async {
        isLoading = true
        errorMessage = nil
        do {
            let resp = try await api.listHotWords(page: page, pageSize: Self.pageSize, search: search)
            hotwords    = resp.hotwords
            total       = resp.total
            currentPage = page
        } catch {
            errorMessage = apiErrorMessage(error)
        }
        isLoading = false
    }

    // MARK: - CRUD

    func create(word: String) async -> Bool {
        do {
            let item = try await api.createHotWord(word: word)
            // 新词插入当前页头部；若当前是搜索状态且新词不匹配搜索词则不插入
            let keyword = loadedSearchText.lowercased()
            if keyword.isEmpty || item.word.lowercased().contains(keyword) {
                hotwords.insert(item, at: 0)
            }
            total += 1
            return true
        } catch {
            errorMessage = apiErrorMessage(error)
            return false
        }
    }

    func update(id: String, word: String) async -> Bool {
        do {
            let updated = try await api.updateHotWord(id: id, word: word)
            if let idx = hotwords.firstIndex(where: { $0.id == id }) {
                hotwords[idx] = updated
            }
            return true
        } catch {
            errorMessage = apiErrorMessage(error)
            return false
        }
    }

    func delete(id: String) async -> Bool {
        do {
            try await api.deleteHotWord(id: id)
            hotwords.removeAll { $0.id == id }
            total = max(0, total - 1)
            // 删完当前页最后一条时自动退到上一页
            if hotwords.isEmpty && currentPage > 1 {
                await goToPage(currentPage - 1)
            }
            return true
        } catch {
            errorMessage = apiErrorMessage(error)
            return false
        }
    }
}

File: lobster-input-win/LobsterInput/Stores/DictionaryStore.cs
using CommunityToolkit.Mvvm.ComponentModel;
using LobsterInput.Models;
using LobsterInput.Services;

namespace LobsterInput.Stores;

public sealed partial class DictionaryStore : ObservableObject
{
    public static DictionaryStore Instance { get; } = new();

    public const int PageSize = 50;
    public const int MaxWordLength = 20;

    [ObservableProperty]
    private List<HotWordItem> _hotWords = new();

    [ObservableProperty]
    private bool _isLoading;

    [ObservableProperty]
    private string? _errorMessage;

    [ObservableProperty]
    private int _currentPage = 1;

    [ObservableProperty]
    private int _total;

    [ObservableProperty]
    private string _searchText = "";

    private string _loadedSearchText = "";

    public int TotalPages => Math.Max(1, (int)Math.Ceiling((double)Total / PageSize));
    public bool HasPrev => CurrentPage > 1;
    public bool HasNext => CurrentPage < TotalPages;

    private readonly ApiClient _api = ApiClient.Instance;

    private DictionaryStore() { }

    public async Task LoadAsync()
    {
        CurrentPage = 1;
        _loadedSearchText = SearchText;
        await FetchPageAsync(1, SearchText);
    }

    public Task GoToPageAsync(int page)
    {
        if (page < 1 || page > TotalPages) return Task.CompletedTask;
        return FetchPageAsync(page, _loadedSearchText);
    }

    public Task PrevPageAsync() => GoToPageAsync(CurrentPage - 1);

    public Task NextPageAsync() => GoToPageAsync(CurrentPage + 1);

    public async Task SearchAsync()
    {
        CurrentPage = 1;
        _loadedSearchText = SearchText;
        await FetchPageAsync(1, SearchText);
    }

    private async Task FetchPageAsync(int page, string search)
    {
        IsLoading = true;
        ErrorMessage = null;

        try
        {
            var resp = await _api.ListHotWordsAsync(page, PageSize, search);
            HotWords = resp.Hotwords;
            Total = ResolveTotal(resp, page);
            CurrentPage = resp.Page > 0 ? resp.Page : page;
            NotifyPagingChanged();
        }
        catch (Exception ex)
        {
            ErrorMessage = ex.ToDisplayMessage();
        }

        IsLoading = false;
    }

    public async Task<bool> AddAsync(string word)
    {
        try
        {
            var item = await _api.CreateHotWordAsync(word);
            if (MatchesLoadedSearch(item.Word))
                HotWords.Insert(0, item);
            Total++;
            OnPropertyChanged(nameof(HotWords));
            NotifyPagingChanged();
            return true;
        }
        catch (Exception ex)
        {
            ErrorMessage = ex.ToDisplayMessage();
            return false;
        }
    }

    public async Task<bool> UpdateAsync(string id, string word)
    {
        try
        {
            var updated = await _api.UpdateHotWordAsync(id, word);
            var idx = HotWords.FindIndex(h => h.Id == id);
            if (idx >= 0)
            {
                HotWords[idx] = updated;
                OnPropertyChanged(nameof(HotWords));
            }
            return true;
        }
        catch (Exception ex)
        {
            ErrorMessage = ex.ToDisplayMessage();
            return false;
        }
    }

    public async Task<bool> DeleteAsync(string id)
    {
        try
        {
            await _api.DeleteHotWordAsync(id);
            HotWords.RemoveAll(h => h.Id == id);
            Total = Math.Max(0, Total - 1);
            if (HotWords.Count == 0 && CurrentPage > 1)
                await GoToPageAsync(CurrentPage - 1);
            else
                OnPropertyChanged(nameof(HotWords));
            NotifyPagingChanged();
            return true;
        }
        catch (Exception ex)
        {
            ErrorMessage = ex.ToDisplayMessage();
            return false;
        }
    }

    private bool MatchesLoadedSearch(string word)
    {
        return string.IsNullOrWhiteSpace(_loadedSearchText)
               || word.Contains(_loadedSearchText, StringComparison.OrdinalIgnoreCase);
    }

    private static int ResolveTotal(HotWordListResponse resp, int page)
    {
        if (resp.Total > 0) return resp.Total;
        var seen = (page - 1) * PageSize + resp.Hotwords.Count;
        return resp.HasMore ? seen + 1 : seen;
    }

    private void NotifyPagingChanged()
    {
        OnPropertyChanged(nameof(TotalPages));
        OnPropertyChanged(nameof(HasPrev));
        OnPropertyChanged(nameof(HasNext));
    }
}
